catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,605 @@
+   *
+   * <p>Does not block
    *
-   * @param source the source to get the {@link CachedSource} for
-   * @return a {@link CachedSource} which contains cached values
+   * @throws IllegalArgumentException if {@code serviceEvent} is {@code null}
    */
-  public CachedSource getCachedSource(Source source) {
-    return runner.getCachedSource(source);
+  @Override
+  public void event(
+      final ServiceEvent serviceEvent,
+      final Map<BundleContext, Collection<ListenerHook.ListenerInfo>> listeners) {
+    notNull(serviceEvent);
+
+    final ServiceReference<?> serviceReference = serviceEvent.getServiceReference();
+    final BundleContext bundleContext = getBundleContext();
+    final Object service = bundleContext.getService(serviceReference);
+
+    if (service instanceof Source) {
+      final Source source = (Source) service;
+      final int serviceEventType = serviceEvent.getType();
+
+      // ServiceEvent.REGISTERED and ServiceEvent.UNREGISTERING are already handled by the
+      // bind and unbind methods, respectively.
+      if (serviceEventType != ServiceEvent.REGISTERED
+          && serviceEventType != ServiceEvent.UNREGISTERING) {
+        final String sourceId = source.getId();
+        final SourceKey sourceKey = new SourceKey(source);
+
+        final Optional<CachedSourceAvailability> cachedSourceStatusOptional =
+            Optional.ofNullable(cache.getIfPresent(sourceKey));
+        if (cachedSourceStatusOptional.isPresent()) {
+          LOGGER.debug(
+              "The source id={} has been updated but its SourceKey is still the same. Cancelling the current availability check of the source, if not completed. Rechecking the availability of the source.",
+              sourceId);
+          final CachedSourceAvailability cachedSourceAvailability =
+              cachedSourceStatusOptional.get();
+          cachedSourceAvailability.cancel();
+          cachedSourceAvailability.recheckOrGetCurrentCheck(source);
+        } else {
+          LOGGER.debug(
+              "Even though the source id={} has already been bound, the SourceKey for the source has been changed. The entry for the old SourceKey will expire.",
+              sourceId);
+          bind(source);
+        }
+      }
+    }
   }
 
   /**
-   * Cancels the {@link SourcePollerRunner} thread that had been previously scheduled to run at
-   * specific intervals. Invoked by the CatalogFramework's blueprint when the framework is
-   * unregistered/uninstalled.
+   * Starts a process to check {@link Source#isAvailable()} of the {@code source}
    *
-   * @param framework unused, but required by blueprint
-   * @param properties unused, but required by blueprint
+   * <p>Does not block
+   *
+   * @throws IllegalArgumentException if {@code source} is {@code null} or if the {@code source} has
+   *     already been bound
    */
-  public void cancel(CatalogFramework framework, Map properties) {
+  public void bind(final Source source) {
+    notNull(source);
```

Comment:
I feel like we sometimes see a bind with null from the system. I think we normally just ignore it rather than throw an exception.

Reply:
N/A after new changes